File: api/internal/crdt/lww_test.go
package crdt

import (
	"testing"
	"time"

	"github.com/vasi1796/doit/internal/hlc"
)

var base = time.Date(2026, 3, 21, 12, 0, 0, 0, time.UTC)

func TestMergeLWW(t *testing.T) {
	tests := []struct {
		name      string
		localVal  string
		localHLC  hlc.Timestamp
		remoteVal string
		remoteHLC hlc.Timestamp
		wantVal   string
	}{
		{
			name:      "remote later by time",
			localVal:  "local",
			localHLC:  hlc.Timestamp{Time: base, Counter: 0},
			remoteVal: "remote",
			remoteHLC: hlc.Timestamp{Time: base.Add(time.Second), Counter: 0},
			wantVal:   "remote",
		},
		{
			name:      "local later by time",
			localVal:  "local",
			localHLC:  hlc.Timestamp{Time: base.Add(time.Second), Counter: 0},
			remoteVal: "remote",
			remoteHLC: hlc.Timestamp{Time: base, Counter: 0},
			wantVal:   "local",
		},
		{
			name:      "same time, remote later by counter",
			localVal:  "local",
			localHLC:  hlc.Timestamp{Time: base, Counter: 1},
			remoteVal: "remote",
			remoteHLC: hlc.Timestamp{Time: base, Counter: 5},
			wantVal:   "remote",
		},
		{
			name:      "same time, local later by counter",
			localVal:  "local",
			localHLC:  hlc.Timestamp{Time: base, Counter: 5},
			remoteVal: "remote",
			remoteHLC: hlc.Timestamp{Time: base, Counter: 1},
			wantVal:   "local",
		},
		{
			name:      "identical timestamps — remote wins (deterministic tiebreak)",
			localVal:  "local",
			localHLC:  hlc.Timestamp{Time: base, Counter: 3},
			remoteVal: "remote",
			remoteHLC: hlc.Timestamp{Time: base, Counter: 3},
			wantVal:   "remote",
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			got, _ := MergeLWW(tc.localVal, tc.localHLC, tc.remoteVal, tc.remoteHLC)
			if got != tc.wantVal {
				t.Errorf("MergeLWW() = %q, want %q", got, tc.wantVal)
			}
		})
	}
}

File: web/src/__tests__/lww.test.ts
import { describe, it, expect } from 'vitest'
import { mergeLWW } from '../crdt/lww'
import type { HLCTimestamp } from '../hlc/hlc'

describe('mergeLWW', () => {
  const cases: {
    name: string
    localVal: string
    localHLC: HLCTimestamp
    remoteVal: string
    remoteHLC: HLCTimestamp
    expectedVal: string
    expectedHLC: HLCTimestamp
  }[] = [
    {
      name: 'remote wins when remote time is later',
      localVal: 'local',
      localHLC: { time: 100, counter: 0 },
      remoteVal: 'remote',
      remoteHLC: { time: 200, counter: 0 },
      expectedVal: 'remote',
      expectedHLC: { time: 200, counter: 0 },
    },
    {
      name: 'local wins when local time is later',
      localVal: 'local',
      localHLC: { time: 200, counter: 0 },
      remoteVal: 'remote',
      remoteHLC: { time: 100, counter: 0 },
      expectedVal: 'local',
      expectedHLC: { time: 200, counter: 0 },
    },
    {
      name: 'remote wins when times equal but remote counter is higher',
      localVal: 'local',
      localHLC: { time: 100, counter: 1 },
      remoteVal: 'remote',
      remoteHLC: { time: 100, counter: 5 },
      expectedVal: 'remote',
      expectedHLC: { time: 100, counter: 5 },
    },
    {
      name: 'local wins when times equal but local counter is higher',
      localVal: 'local',
      localHLC: { time: 100, counter: 5 },
      remoteVal: 'remote',
      remoteHLC: { time: 100, counter: 1 },
      expectedVal: 'local',
      expectedHLC: { time: 100, counter: 5 },
    },
    {
      name: 'remote wins on exact tie (deterministic tiebreaker)',
      localVal: 'local',
      localHLC: { time: 100, counter: 3 },
      remoteVal: 'remote',
      remoteHLC: { time: 100, counter: 3 },
      expectedVal: 'remote',
      expectedHLC: { time: 100, counter: 3 },
    },
  ]

  for (const tc of cases) {
    it(tc.name, () => {
      const [val, hlc] = mergeLWW(tc.localVal, tc.localHLC, tc.remoteVal, tc.remoteHLC)
      expect(val).toBe(tc.expectedVal)
      expect(hlc).toEqual(tc.expectedHLC)
    })
  }

  it('works with non-string types (numbers)', () => {
    const [val, hlc] = mergeLWW(10, { time: 50, counter: 0 }, 20, { time: 100, counter: 0 })
    expect(val).toBe(20)
    expect(hlc).toEqual({ time: 100, counter: 0 })
  })
})
